src/test/kotlin/com/safehill/kcrypto/CipherTests.kt

```diff
@@ -1,110 +1,133 @@
+            assertEquals(code1.length, 6)
+            assertEquals(code1, code2)
 
-        val (code5, _) = SafehillCypher.generateOTPCode(
-            secret = secret,
-            digits = 6,
-            timeStepInSeconds = 1
-        )
-        assertNotEquals(code4, code5)
+            val safehillOTP = SafehillOTP(digits = 6, validDuration = 1.seconds)
+            val (code3, _) = safehillOTP.generateCode(secret)
+            val isValid = safehillOTP.isValid(secret = secret, code3)
+            assert(isValid)
 
-        val (code6, _) = SafehillCypher.generateOTPCode(
-            secret = secret,
-            digits = 6,
-            timeStepInSeconds = 2
-        )
-        assertNotEquals(code5, code6)
 
-        val (code7, _) = SafehillCypher.generateOTPCode(
-            secret = secret,
-            digits = 6,
-            timeStepInSeconds = 2
-        )
-        assertEquals(code6, code7)
+            val (code5, _) = SafehillOTP(digits = 6, validDuration = 1.seconds).generateCode(secret)
+            delay(0.5.seconds)
+            assert(SafehillOTP(digits = 6, validDuration = 1.seconds).isValid(secret, code5))
+            delay(0.6.seconds)
+            assertFalse(SafehillOTP(digits = 6, validDuration = 1.seconds).isValid(secret, code5))
 
-        Thread.sleep(1100)
 
-        val (code8, valid8) = SafehillCypher.generateOTPCode(
-            secret = secret,
-            digits = 6,
-            timeStepInSeconds = 2
-        )
-        Thread.sleep(max(0, valid8 - 100))
-        val (code9, valid9) = SafehillCypher.generateOTPCode(
-            secret = secret,
-            digits = 6,
-            timeStepInSeconds = 2
-        )
-        assertEquals(code8, code9)
-        Thread.sleep(valid9 + 1)
-        val (code10, _) = SafehillCypher.generateOTPCode(
-            secret = secret,
-            digits = 6,
-            timeStepInSeconds = 2
-        )
-        assertNotEquals(code9, code10)
+            val (code8, valid8) = SafehillOTP(digits = 6, validDuration = 2.seconds).generateCode(secret)
+            delay((valid8 - 100.milliseconds).coerceAtLeast(0.milliseconds))
+            assert(SafehillOTP(digits = 6, validDuration = 2.seconds).isValid(secret, code8))
+
+
+
+            val (code10, _) = SafehillOTP(digits = 6, validDuration = 2.seconds).generateCode(secret)
+
+            val newSecret = SafehillCypher.generateRandomIV()
+            val isValid1 = SafehillOTP(digits = 6, validDuration = 2.seconds).isValid(
+                secret = newSecret, code = code10
+            )
+            assertFalse(isValid1)
+
+            val (code11, _) = SafehillOTP(digits = 6, validDuration = 2.seconds).generateCode(newSecret)
+
+            delay(0.99.seconds)
+
+            assert(SafehillOTP(digits = 6, validDuration = 2.seconds).isValid(newSecret, code11))
+        }
+
+    }
+
+    @Test
+    fun `test otp generated is valid for minimum of half the time step`() {
```

Comment:
Can you make sure the unit tests cover generation and verification across all possible cases?
Think of where generation and verification lay in the timeline and intervals you drew on the board yesterday:
- both generation and verification in first interval
- generation in first and verification in second
- generation in first and verification after second
- etc.